File: ch02/item11/CppDestructor/CriticalSection.cpp
#include "pch.h"
#include <iostream>
#include "CriticalSection.h"


CriticalSection::CriticalSection()
{
	EnterCriticalSection();
}


CriticalSection::~CriticalSection()
{
	ExitCriticalSection();
}

void CriticalSection::ExitCriticalSection()
{
	std::cout << "ExitCriticalsection()\n";
}
void CriticalSection::EnterCriticalSection()
{
	std::cout << "EnterCriticalsection()\n";
}

File: ch02/item11/CSharpFinalizer/CriticalSection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharpFinalizer
{
    class CriticalSection
    {
        public CriticalSection()
        {
            EnterCriticalSection();
        }

        ~CriticalSection()
        {
            ExitCriticalSection();
        }

        private void EnterCriticalSection()
        {
            Console.WriteLine("CriticalSection.EnterCriticalSection()");
        }

        private void ExitCriticalSection()
        {
            Console.WriteLine("CriticalSection.ExitCriticalSection()");
        }

    }
}
